### DISCUSSION TITLE
unable to use @babel/plugin-proposal-decorators with standalone

### DISCUSSION THREAD

**Discussion by ChristopherHButler** [babel/babel#12405] [Q&A] [answered]
I am trying to use babel standalone inside a react app to transpile Angular TypeScript. Because Angular uses TypeScript and decorators I need to add plugins to support decorators and class properties but I get errors when using the plugins.

Options: 
```
const TS_OPTIONS = {
  presets: [
    'typescript',
    ['es2017', { 'modules': false }],
  ],
  plugins: [
    ['plugin-proposal-decorators', { 'legacy': true }],
    ['plugin-proposal-class-properties', { 'loose': true }],
    'transform-es2015-modules-commonjs',
  ],
};
```

My transpile function (simplified): 
```
export default function transpile(myCode) {
  const { code } = Babel.transform(myCode, TS_OPTIONS);

  return myCode;
}

```

No matter how I write the plugins it does not work. I keep getting an error along the lines of

```
Error: Invalid plugin specified in Babel options: "plugin-proposal-decorators"
```
I have also tried following the instructions [here](https://babeljs.io/docs/en/babel-plugin-proposal-decorators) but it seems @babel does not work with babel standalone. 

Can someone explain how I can use these plugins with babel standalone?

**Comment by nicolo-ribaudo** [comment, marked as answer]
Are you using Babel 6? That plugin was introduced in Babel 7.

**Reply by ChristopherHButler** [reply]
No, I am using 7 from a cdn: https://cdnjs.cloudflare.com/ajax/libs/babel-standalone/7.0.0-beta.3/babel.min.js

**Reply by nicolo-ribaudo** [reply]
Ok, that it a very old beta of Babel 7 and it's possible that we didn't add support for decorators yet.

cdnjs doesn't support Babel 7 (except for that old beta version), but you can use a different CDN: https://unpkg.com/@babel/standalone@7.12.9/babel.min.js

Keep in mind that there were different breaking changes, so make sure to check https://babeljs.io/docs/en/v7-migration

**Reply by ChristopherHButler** [reply]
Hey thanks nicolo,
So I was able to upgrade to the newer version and get these plugins "working" however I'm still experiencing my original issue with regards to transpiling Angular TypeScript which I believe is related to the decorators proposal plugin.  The error I get is:

```
Uncaught Error: Template parse errors:
'search-bar' is not a known element:
1. If 'search-bar' is an Angular component, then verify that it is part of this module.
2. If 'search-bar' is a Web Component then add 'CUSTOM_ELEMENTS_SCHEMA' to the '@NgModule.schemas' of this component to suppress this message.
```  
However, this is only if I try to import an Angular component and try to use it's selector in another component's template like this:

Search Bar component
```
// SearchBar
const { Component, NgModule, Output, EventEmitter } = ng.core;
const { CommonModule } = ng.common;

// Component Template (decorator)
@Component({
  selector: 'search-bar',
  template: `<div>Search Bar</div>`,
})
@NgModule({
  declarations: [SearchBar],
  exports: [SearchBar]
})
class SearchBar {}

export default SearchBar;
```

App component. This may look a little strange that there is no path to the SearchBar component but this is specific to my app and does not cause any errors, it's just how it works. 
```
const { Component, NgModule, VERSION } = ng.core;
const { CommonModule } = ng.common;

import SearchBar from 'SearchBar.ts';

@NgModule({
  imports: [
    CommonModule,
    SearchBar,
  ],
  declarations: [AppComponent, SearchBar],
})
@Component({
  selector: 'my-app',
  template: `<search-bar></search-bar>`
})
export default class AppComponent {

}

```

My current version of babel:
https://unpkg.com/@babel/standalone@7.12.9/babel.min.js

My babel config:
```
const TS_OPTIONS = {
  presets: [
    ['typescript', { allExtensions: true }],
    ['es2017', { 'modules': false }],
  ],
  plugins: [
    ["proposal-decorators", { "legacy": true }],
    ["proposal-class-properties", { "loose" : true }],
    'transform-modules-commonjs',
  ],
};
```
I understand you may not be familiar with Angular but I am more interested to find out if this is a babel issue.

**Reply by ChristopherHButler** [reply]
I was able to solve this. Answer here for anyone interested: [https://stackoverflow.com/questions/65040045/angular-uncaught-error-template-parse-errors-is-not-a-known-element/65043707#65043707](https://stackoverflow.com/questions/65040045/angular-uncaught-error-template-parse-errors-is-not-a-known-element/65043707#65043707). Thanks again @nicolo-ribaudo
